packages/perseus/src/widgets/matrix/matrix-validator.ts

```diff
@@ -0,0 +1,85 @@
+import _ from "underscore";
+
+import KhanAnswerTypes from "../../util/answer-types";
+
+import {getMatrixSize} from "./matrix";
+
+import type {PerseusStrings} from "../../strings";
+import type {PerseusScore} from "../../types";
+import type {
+    PerseusMatrixRubric,
+    PerseusMatrixUserInput,
+} from "../../validation.types";
+
+function matrixValidator(
+    state: PerseusMatrixUserInput,
+    rubric: PerseusMatrixRubric,
+    strings: PerseusStrings,
+): PerseusScore {
+    const solution = rubric.answers;
+    const supplied = state.answers;
+    const solutionSize = getMatrixSize(solution);
+    const suppliedSize = getMatrixSize(supplied);
+
+    const incorrectSize =
+        solutionSize[0] !== suppliedSize[0] ||
+        solutionSize[1] !== suppliedSize[1];
+
+    const createValidator = KhanAnswerTypes.number.createValidatorFunctional;
+    let message = null;
+    let hasEmptyCell = false;
+    let incorrect = false;
+    _(suppliedSize[0]).times((row) => {
+        _(suppliedSize[1]).times((col) => {
+            if (
+                supplied[row][col] == null ||
+                supplied[row][col].toString().length === 0
+            ) {
+                hasEmptyCell = true;
+            }
+            if (!incorrectSize) {
+                const validator = createValidator(
+                    // @ts-expect-error - TS2345 - Argument of type 'number' is not assignable to parameter of type 'string'.
+                    solution[row][col],
+                    {
+                        simplify: true,
+                    },
+                    strings,
+                );
+                const result = validator(supplied[row][col]);
+                if (result.message) {
+                    // @ts-expect-error - TS2322 - Type 'string' is not assignable to type 'null'.
+                    message = result.message;
+                }
+                if (!result.correct) {
+                    incorrect = true;
+                }
+            }
+        });
+    });
+
+    if (hasEmptyCell) {
+        return {
+            type: "invalid",
+            message: strings.fillAllCells,
+        };
+    }
+
+    if (incorrectSize) {
```

Comment:
nit: For efficiency it would make more sense to return this error at the top of the function instead of placing at the bottom for after the computation has run. Even though I doubt our complexity will get more than `n^2`. Just food for thought 😉

Reply:
That's a good point. I didn't write this code, but I'm wondering if they did this to give...

``` TS
    if (hasEmptyCell) {
        return {
            type: "invalid",
            message: strings.fillAllCells,
        };
    }
```

...a chance to go first. They needed `incorrectSize` in that loop, but they needed to run the loop to get `hasEmptyCell`. Invalid scores don't count against learners, but incorrect scores do; so they waited to return the incorrect score until after they had a chance to send the invalid score.

I'm trying to not be too opinionated about the code I'm moving though since we're having to do this for all widgets, so I'm going to leave it as-is.